### DISCUSSION TITLE
Different length for responded data and includes for APIv2

### DISCUSSION THREAD

**Discussion by AndrewQiu9748** [tweepy/tweepy#1799] [Questions] [answered]
Hi, I use client.search_all_tweets in APIv2, and I also need the username from ```includes``` in addition to ```data```. However, I found the length is different, which makes it difficult for me to combine them. Here is the code

```
for tweet in tweepy.Paginator(client.search_all_tweets, query=query,
                              expansions=['author_id'],tweet_fields=['author_id','context_annotations', 'created_at', 'geo', public_metrics','referenced_tweets'], start_time=start_time,
                              end_time=end_time, max_results=100):
    id_temp = [tweet.id for tweet in tweet.data]
    author_name_temp = [name for name in tweet.includes["users"]]
    print(len(tweet.data))
    print(len(tweet.includes["users"]))
    time.sleep(1)
```

and responses:
```99
99
100
99
100
99```

I don't know why sometimes the numbers of includes are less than the numbers of data. When I create dataframe it shows "Arrays Must be All Same Length" as the numbers do not match, and I do not even know which one is missing. 

Can anyone help me? Thanks!

**Comment by Harmon758** [comment, marked as answer]
Are there multiple Tweets from the same user?

For code block usage, see https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks.

**Reply by AndrewQiu9748** [reply]
Yes, you are right. When one user post multiple tweets, there will be only one user in the ```includes``` but multiple ```data```.

So how can I match them in each response? Here is my code:

```
for tweet in tweepy.Paginator(client.search_all_tweets, query=query,
                              expansions=['author_id'],tweet_fields=['author_id','context_annotations', 'created_at', 'geo', 'public_metrics','referenced_tweets'], start_time=start_time,
                              end_time=end_time, max_results=100, limit =1):
    id_temp = [tweet.id for tweet in tweet.data]
    author_name_temp = [name for name in tweet.includes["users"]]
    author_id = [tweet.author_id for tweet in tweet.data]
    text_temp = [tweet.text for tweet in tweet.data]
```
It seems that the response data of ```includes.users``` have ids, i.e., <User id=1301468160525324288 name=Sandelle Bañaga Andallo username=SandelleA>, but only the username is displayed when I use it. Otherwise maybe I can use the author_id from ```data``` to match the user id from ```includes[users"]```.

Can anyone help?
Many thanks!!

**Reply by Harmon758** [reply]
Yes, you can match them by ID. I'm not sure what you mean by "only the username is displayed when I use it".

**Reply by 2jjk** [reply]
> Yes, you are right. When one user post multiple tweets, there will be only one user in the `includes` but multiple `data`.
> 
> So how can I match them in each response? Here is my code:
> 
> ```
> for tweet in tweepy.Paginator(client.search_all_tweets, query=query,
>                               expansions=['author_id'],tweet_fields=['author_id','context_annotations', 'created_at', 'geo', 'public_metrics','referenced_tweets'], start_time=start_time,
>                               end_time=end_time, max_results=100, limit =1):
>     id_temp = [tweet.id for tweet in tweet.data]
>     author_name_temp = [name for name in tweet.includes["users"]]
>     author_id = [tweet.author_id for tweet in tweet.data]
>     text_temp = [tweet.text for tweet in tweet.data]
> ```
> 
> It seems that the response data of `includes.users` have ids, i.e., <User id=1301468160525324288 name=Sandelle Bañaga Andallo username=SandelleA>, but only the username is displayed when I use it. Otherwise maybe I can use the author_id from `data` to match the user id from `includes[users"]`.
> 
> Can anyone help? Many thanks!!

I'm using a similar code here, but are you getting the public_metrics in the Response object? tweet.public_metrics worked when there was no pagination, and I have it included in the tweet_fields. It doesn't work with tweepy.Paginator (tried tweet.data.public_metrics as well). It isn't included in the Response object. I'm using search_recent_tweets, so is that it? Can it only work with the full search API?

**Reply by Harmon758** [reply]
How are you attempting to access the metrics?

**Reply by AndrewQiu9748** [reply]
> Yes, you can match them by ID. I'm not sure what you mean by "only the username is displayed when I use it".

I can't get Userid from <User id=1301468160525324288 name=Sandelle Bañaga Andallo username=SandelleA> from ```inclueds.users```, because when I use ```print(includes.users)```, it only displays username without id or name.

Anyway, now I use client.get_users with author_id and successfully get the username.

**Reply by AndrewQiu9748** [reply]
> > Yes, you are right. When one user post multiple tweets, there will be only one user in the `includes` but multiple `data`.
> > So how can I match them in each response? Here is my code:
> > ```
> > for tweet in tweepy.Paginator(client.search_all_tweets, query=query,
> >                               expansions=['author_id'],tweet_fields=['author_id','context_annotations', 'created_at', 'geo', 'public_metrics','referenced_tweets'], start_time=start_time,
> >                               end_time=end_time, max_results=100, limit =1):
> >     id_temp = [tweet.id for tweet in tweet.data]
> >     author_name_temp = [name for name in tweet.includes["users"]]
> >     author_id = [tweet.author_id for tweet in tweet.data]
> >     text_temp = [tweet.text for tweet in tweet.data]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It seems that the response data of `includes.users` have ids, i.e., <User id=1301468160525324288 name=Sandelle Bañaga Andallo username=SandelleA>, but only the username is displayed when I use it. Otherwise maybe I can use the author_id from `data` to match the user id from `includes[users"]`.
> > Can anyone help? Many thanks!!
> 
> I'm using a similar code here, but are you getting the public_metrics in the Response object? tweet.public_metrics worked when there was no pagination, and I have it included in the tweet_fields. It doesn't work with tweepy.Paginator (tried tweet.data.public_metrics as well). It isn't included in the Response object. I'm using search_recent_tweets, so is that it? Can it only work with the full search API?

It does work with tweepy.Paginator when I am using full search API, not sure if it works with search_recent_tweets.

**Reply by Harmon758** [reply]
> <User id=1301468160525324288 name=Sandelle Bañaga Andallo username=SandelleA>

1301468160525324288 is the user ID here.

**Reply by AndrewQiu9748** [reply]
> > <User id=1301468160525324288 name=Sandelle Bañaga Andallo username=SandelleA>
> 
> 1301468160525324288 is the user ID here.

Yes, I know. But I cannot get it out from ```includes.users```
The id is missing when I print it:

```
tweet.includes["users"][1]
<User id=253483318 name=Electra 🐝 #JohnsonOut #JohnsonMustGo username=electra316>
print(tweet.includes["users"][1])
electra316
```

**Reply by Harmon758** [reply]
The former is the string representation of the object.
The latter is the `User` object cast to a string.
You can access the ID from the object itself by attribute or key.
See the [FAQ section about this in the documentation](https://tweepy.readthedocs.io/en/latest/faq.html#why-am-i-not-getting-expansions-or-fields-data-with-api-v2-using-client).

**Reply by AndrewQiu9748** [reply]
> The former is the string representation of the object. The latter is the `User` object cast to a string. You can access the ID from the object itself by attribute or key. See the [FAQ section about this in the documentation](https://tweepy.readthedocs.io/en/latest/faq.html#why-am-i-not-getting-expansions-or-fields-data-with-api-v2-using-client).

Thanks! Problem solved!!
